### PR TITLE
Implement web-hooks

### PR DESCRIPTION
**Pull request by kiootic**
ref #961

### CODE REVIEW

**Comment by kiootic** [comment]
Let's put this PR on hold until I finish the web-hook part then.

**Comment by kiootic** [comment]
@louischan-oursky the implementation should be completed, please review.

File: pkg/auth/dependency/hook/provider_impl.go [hunk 1]
```diff
@@ -63,6 +71,11 @@ func (provider *providerImpl) DispatchEvent(payload event.Payload, user *model.U
 }
 
 func (provider *providerImpl) WillCommitTx() error {
+	err := provider.syncUserIfNeeded()
```
Comment (louischan-oursky) [inline pkg/auth/dependency/hook/provider_impl.go]: Maybe rename it to `addSyncUserEventIfNeeded`

File: pkg/auth/event/event.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	Type       Type        `json:"type"`
+	Payload    interface{} `json:"payload"`
+	Context    Context     `json:"context"`
+}
+
+type Context struct {
+	Timestamp   int64   `json:"timestamp"`
+	RequestID   *string `json:"request_id"`
+	UserID      *string `json:"user_id"`
+	PrincipalID *string `json:"identity_id"`
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/event/event.go]: These fields are nullable.

File: pkg/auth/event/hook_response.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+type HookResponse struct {
+	IsAllowed bool       `json:"is_allowed"`
+	Reason    string     `json:"reason"`
+	Mutations *Mutations `json:"mutations"`
+}
+
+func (resp HookResponse) Validate() error {
+	if resp.IsAllowed {
+		if resp.Mutations != nil {
+			return skyerr.NewInvalidArgument("mutations must not exist", []string{"mutations"})
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/event/hook_response.go]: If is it disallowed, the operation would be cancelled, and the transaction would be rollbacked. If we allow mutations, then a new transaction would be started to specifically perform the mutations. I'd prefer the handler to call API directly.

File: pkg/auth/handler/sso/customtoken.go [hunk 4]
```diff
@@ -240,6 +244,31 @@ func (h CustomTokenLoginHandler) Handle(req interface{}) (resp interface{}, err
 
 	// TODO: check disable
 
+	// Populate the activity time to user
+	now := timeNow()
+	info.LastLoginAt = &now
+	info.LastSeenAt = &now
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/handler/sso/customtoken.go]: I'll add a comment to clarify the intention.

File: pkg/auth/handler/sso/auth_handler.go [hunk 5]
```diff
@@ -130,63 +130,73 @@ func (h AuthHandler) DecodeRequest(request *http.Request) (handler.RequestPayloa
...
+	if err == nil {
+		h.HookProvider.DidCommitTx()
+	}
+}
+
+func (h AuthHandler) Handle(w http.ResponseWriter, r *http.Request) (success bool) {
 	var ok interface{}
-	var err error
 	var oauthAuthInfo sso.AuthInfo
+	success = false
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/handler/sso/auth_handler.go]: The function is already handling the errors itself, and lifting its error handling logic out would be a complicated refactoring. If you prefer returning and error, I could `return authHandlerError{}` in the function instead.

File: pkg/auth/dependency/hook/deliverer.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package hook
+
+import (
+	"github.com/skygeario/skygear-server/pkg/auth/event"
+	"github.com/skygeario/skygear-server/pkg/auth/model"
+)
+
+type Deliverer interface {
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/dependency/hook/deliverer.go]: The wording used in spec is 'delivering events to handler`, so the noun form of 'deliver' would make more sense.

File: pkg/auth/dependency/time/mock_provider.go [hunk 7]
```diff
@@ -4,10 +4,15 @@ import "time"
...
 type MockProvider struct {
 	TimeNowUTC time.Time
+	TimeNow    time.Time
 }
 
 func (provider MockProvider) NowUTC() time.Time {
 	return provider.TimeNowUTC
 }
 
+func (provider MockProvider) Now() time.Time {
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/dependency/time/mock_provider.go]: `Now().UTC()` would strip the monotonic clock reading from the time. So for measuring timeout purpose, `Now()` would be used instead.

File: pkg/core/hash/hmac_sha256.go [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package hash
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+)
+
+// HMACSHA256 returns the HMAC-SHA256 code of body using secret as key.
+func HMACSHA256(body []byte, secret []byte) string {
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/core/hash/hmac_sha256.go]: I'll suggest renaming the `hash` package into `crypto` package instead.

File: pkg/auth/event/event.go [hunk 9]
```diff
@@ -8,6 +8,12 @@ type Payload interface {
 	Version() int32
 }
 
+// UserAwarePayload represents event payload that can apply mutations on its own user object
+type UserAwarePayload interface {
+	Payload
+	ApplyingMutations(Mutations) UserAwarePayload
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/event/event.go]: It's a naming convention borrowed from Swift: `ApplyMutations` sounds like it is mutating the payload, and `ApplyingMutations` would return new payload and leave the current one unchanged.

File: pkg/auth/event/user_update.go [hunk 10]
```diff
@@ -41,3 +41,19 @@ func (UserUpdateEvent) BeforeEventType() Type {
 func (UserUpdateEvent) AfterEventType() Type {
 	return AfterUserUpdate
 }
+
+func (event UserUpdateEvent) ApplyingMutations(mutations Mutations) UserAwarePayload {
+	// user object in this event is a snapshot before operation, so mutations are not applied
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/event/user_update.go]: No, the mutation is not applied to user object. It is applied to a copy of the event and return it.

File: pkg/auth/dependency/hook/deliverer_mock.go [hunk 11]
```diff
@@ -16,18 +17,25 @@ type mockDelivererBeforeEvent struct {
 }
 
 func newMockDeliverer() *mockDeliverer {
-	return &mockDeliverer{}
+	return &mockDeliverer{
+		BeforeEvents: []mockDelivererBeforeEvent{},
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/dependency/hook/deliverer_mock.go]: For more consistent test cases: `So(x, ShouldResemble, []Item{})` fails if x is `nil`.

File: pkg/auth/dependency/hook/store_impl.go [hunk 12]
```diff
@@ -10,12 +10,14 @@ import (
...
 	sqlBuilder  db.SQLBuilder
 	sqlExecutor db.SQLExecutor
+	events      []*event.Event
 }
 
 func NewStore(builder db.SQLBuilder, executor db.SQLExecutor) Store {
 	return &storeImpl{
 		sqlBuilder:  builder,
 		sqlExecutor: executor,
+		events:      []*event.Event{},
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/dependency/hook/store_impl.go]: Just a habit, I'll remove it.

File: pkg/auth/handler/sso/auth_handler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/handler/sso/auth_handler.go]: Fair enough.

File: pkg/core/hash/hmac_sha256.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/core/hash/hmac_sha256.go]: Fine. Let's follow your suggestion.

File: pkg/auth/event/event.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/event/event.go]: The naming convention in Golang is `WithXXX`. See https://golang.org/pkg/context/#WithValue

File: pkg/auth/event/mutation.go [hunk 13]
```diff
@@ -1,9 +1,42 @@
...
+	if newMutations.VerifyInfo != nil {
+		mutations.VerifyInfo = newMutations.VerifyInfo
+	}
+	if newMutations.Metadata != nil {
+		mutations.Metadata = newMutations.Metadata
+	}
+	return mutations
+}
+
+func (mutations Mutations) ApplyingToUser(user *model.User) *model.User {
```
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/event/mutation.go]: Here we cannot use `WithXXX` because the type of the receiver itself is not the return type. What about `ApplyToClonedUser`?

File: pkg/auth/dependency/hook/deliverer_mock.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/dependency/hook/deliverer_mock.go]: I would write `So(len(x), ShouldEqual, 0)`

File: pkg/auth/dependency/hook/deliverer_impl.go [hunk 14]
```diff
@@ -3,34 +3,38 @@ package hook
...
-	UserConfig   *config.HookUserConfiguration
-	AppConfig    *config.HookAppConfiguration
-	TimeProvider time.Provider
-	Mutator      Mutator
+	Hooks         *[]config.Hook
+	UserConfig    *config.HookUserConfiguration
+	AppConfig     *config.HookAppConfiguration
+	TimeProvider  time.Provider
+	Mutator       Mutator
+	NewHTTPClient func() gohttp.Client
```
Comment (louischan-oursky) [inline pkg/auth/dependency/hook/deliverer_impl.go]: Why not just `HTTPClient *gohttp.Client`?

File: pkg/auth/event/event.go [hunk 15]
```diff
@@ -37,7 +37,7 @@ type Event struct {
 
 func newEvent(seqNo int64, payload Payload, context Context) *Event {
 	return &Event{
-		ID:      uuid.New(),
+		ID:      fmt.Sprintf("%08x", seqNo),
```
Comment (louischan-oursky) [inline pkg/auth/event/event.go]: Why do we need to pad leading zeros?

File: pkg/auth/handler/sso/auth_handler.go [hunk 16]
```diff
@@ -139,73 +139,72 @@ func (authHandlerError) Error() string {
...
+	if err == nil {
+		ok, err = h.handle(oauthAuthInfo, *state)
 	}
-	ok, err = h.handle(oauthAuthInfo, *state)
-	return
+
+	var handlerError error
+	if err == nil {
+		handlerError = authHandlerError{}
+	}
```
Comment (louischan-oursky) [inline pkg/auth/handler/sso/auth_handler.go]: The condition seems wrong.

File: pkg/auth/event/mutation.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/event/mutation.go]: I'll just name it `ApplyToUser` and pass in a copy of user instead.

File: pkg/auth/event/event.go (hunk 15 above)
Comment (kiootic) [reply pkg/auth/event/event.go]: - All ID would have same length
- Can be indexed lexically with same order as sequence number

File: pkg/auth/handler/sso/auth_handler.go (hunk 16 above)
Comment (kiootic) [reply pkg/auth/handler/sso/auth_handler.go]: I'll drop this change if original code is ok.

File: pkg/auth/event/event.go (hunk 15 above)
Comment (louischan-oursky) [reply pkg/auth/event/event.go]:
```typescript
Math.pow(2, 64) === 18446744073709552000;
"18446744073709552000".length === 20;
```

Should we make width 20 instead of 8?

Comment (kiootic) [reply pkg/auth/event/event.go]: It should be 16 characters, hex encoding is used.

Comment (louischan-oursky) [reply pkg/auth/event/event.go]: Why not just use decimal?

Comment (kiootic) [reply pkg/auth/event/event.go]: It's shorter.

Comment (kiootic) [reply pkg/auth/event/event.go]: It's shorter.
